cdap-security/src/main/java/io/cdap/cdap/security/tools/AccessTokenGeneratorService.java

```diff
@@ -0,0 +1,216 @@
+      new IOModule(),
+      new AbstractModule() {
+        @Override
+        protected void configure() {
+          // We don't do anything with metrics, so just bind to the no-op implementation
+          bind(MetricsCollectionService.class).to(NoOpMetricsCollectionService.class);
+        }
+      },
+      CoreSecurityRuntimeModule.getDistributedModule(cConf));
+    CommonNettyHttpServiceFactory factory = injector.getInstance(CommonNettyHttpServiceFactory.class);
+    handler = injector.getInstance(TokenGeneratorHttpHandler.class);
+    NettyHttpService.Builder builder = factory.builder("accesstoken.generator")
+      .setHost(address)
+      .setPort(port)
+      .setExecThreadPoolSize(execThreadPoolSize)
+      .setWorkerThreadPoolSize(workerThreadPoolSize)
+      .setExceptionHandler(new HttpExceptionHandler())
+      .setHttpHandlers(handler);
+    if (cConf.getBoolean(Constants.Security.SSL.INTERNAL_ENABLED)) {
+      SConfiguration sConf = injector.getInstance(SConfiguration.class);
+      new HttpsEnabler().configureKeyStore(cConf, sConf).enable(builder);
+    }
+    httpService = builder.build();
+  }
+
+  @VisibleForTesting
+  InetSocketAddress getBindAddress() {
+    return httpService.getBindAddress();
+  }
+
+  @VisibleForTesting
+  TokenManager getTokenManager() {
+    return handler.tokenManager;
+  }
+
+  @VisibleForTesting
+  Codec<AccessToken> getTokenCodec() {
+    return handler.tokenCodec;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    httpService.start();
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    httpService.stop();
+  }
+
+  @Path(Constants.Gateway.INTERNAL_API_VERSION_3 + "/accesstoken")
+  public static class TokenGeneratorHttpHandler extends AbstractHttpHandler {
+    private final CConfiguration cConf;
+    final TokenManager tokenManager;
+    final Codec<AccessToken> tokenCodec;
+
+    @Inject
+    public TokenGeneratorHttpHandler(CConfiguration cConf, TokenManager tokenManager,
+                                     Codec<AccessToken> tokenCodec) {
+      this.cConf = cConf;
+      this.tokenManager = tokenManager;
+      this.tokenCodec = tokenCodec;
+    }
+
+    @Override
+    public void init(HandlerContext context) {
+      tokenManager.startUp();
+    }
+
+    @Override
+    public void destroy(HandlerContext context) {
+      tokenManager.stopAndWait();
+    }
+
+    @GET
+    @Path("/")
+    public void getToken(FullHttpRequest request, HttpResponder responder,
+                         @QueryParam("username") String username,
+                         @QueryParam("identifierType") @DefaultValue("INTERNAL") String identifierType)
+      throws IOException, BadRequestException {
+      UserIdentity.IdentifierType type;
+      try {
+        type = UserIdentity.IdentifierType.valueOf(identifierType);
+      } catch (IllegalArgumentException e) {
+        throw new BadRequestException("Invalid identifierType. Valid values: INTERNAL, EXTERNAL");
```

Comment:
The valid values should come from `UserIdentity.IdentifierType.values()`

Reply:
fixed